services/motion/builtin/replanner.go

```diff
@@ -0,0 +1,39 @@
+package builtin
+
+import (
+	"context"
+	"time"
+)
+
+type replanResponse struct {
+	err    error
+	replan bool
+}
+
+type replanner struct {
+	period       time.Duration
+	fn           func(ctx context.Context) replanResponse
+	responseChan chan replanResponse
+}
+
+func (r *replanner) startPolling(ctx context.Context) {
+	ticker := time.NewTicker(r.period)
+	defer ticker.Stop()
+	for {
+		// this ensures that if the context is cancelled we always return early at the top of the loop
+		if err := ctx.Err(); err != nil {
+			return
+		}
+
+		select {
+		case <-ctx.Done():
```

Comment:
this seems redundant, in my understanding 
```
if ctx.Err() != nil {
    return 
}
```
is the same check

Reply:
The above check is necessary too 
> this ensures that if the context is cancelled we always return early at the top of the loop
The select statement is nondeterministic in what it chooses so if at the start of the loop multiple cases are true there is no telling what will happen.  We want to catch the case where it is cancelled both before and during the blocking call